include/lsst/afw/table/io/FitsSchemaInputMapper.h

```diff
@@ -0,0 +1,160 @@
+#define AFW_TABLE_IO_FitsSchemaInputMapper_h_INCLUDED
+
+#include "lsst/afw/fits.h"
+#include "lsst/afw/table/Schema.h"
+#include "lsst/afw/table/io/InputArchive.h"
+#include "lsst/afw/table/BaseRecord.h"
+
+namespace lsst { namespace afw { namespace table { namespace io {
+
+class FitsColumnReader {
+public:
+
+    FitsColumnReader() {}
+
+    // Neither copyable nor moveable.
+    FitsColumnReader(FitsColumnReader const &) = delete;
+    FitsColumnReader(FitsColumnReader &&) = delete;
+    FitsColumnReader & operator=(FitsColumnReader const &) = delete;
+    FitsColumnReader & operator=(FitsColumnReader &&) = delete;
+
+    virtual void readCell(
+        BaseRecord & record,
+        std::size_t row,
+        fits::Fits & fits,
+        PTR(InputArchive) const & archive
+    ) const = 0;
+
+    virtual ~FitsColumnReader() {}
+
+};
+
+/**
+ *  A structure that describes a field as a collection of related strings read from the FITS header.
+ */
+struct FitsSchemaItem {
+    int column;          // column number (0-indexed); -1 for Flag fields
+    int bit;             // flag bit number (0-indexed); -1 for non-Flag fields
+    std::string ttype;   // name of the field (from TTYPE keys)
+    std::string tform;   // FITS column format code (from TFORM keys)
+    std::string tccls;   // which field class to use (from our own TCCLS keys)
+    std::string tunit;   // field units (from TUNIT keys)
+    std::string doc;     // field docs (from comments on TTYPE keys)
+
+    explicit FitsSchemaItem(int column_, int bit_) : column(column_), bit(bit_) {}
+};
+
+
+/**
+ *  A class that describes a mapping from a FITS binary table and an afw::table Schema.
+ *
+ *  A FitsSchemaInputMapper is created every time a FITS binary table is read into an afw::table
+ *  catalog, allowing limited customization of the mapping between on-disk FITS table columns
+ *  an in-memory fields by subclasses of BaseTable.
+ *
+ *  The object is constructed from a daf::base::PropertyList that represents the FITS header,
+ *  which is used to populate a custom container of FitsSchemaItems.  These can then be retrieved
+ *  by name or column number via the find() methods, allowing the user to create custom readers
+ *  for columns or groups of columns via addColumnReader().  They can also be removed from the
+ *  "regular" fields via the erase() method.  Those regular fields are filled in by the finalize()
+ *  method, which automatically generates mappings for any FitsSchemaItems that have not been
+ *  removed by calls to erase().  Once finalize() has been called, readRecord() may be called
+ *  repeatedly to read FITS rows into record objects according to the mapping that has been
+ *  defined.
+ */
+class FitsSchemaInputMapper {
+public:
+
+    typedef FitsSchemaItem Item;
+
+    /// Construct a mapper from a PropertyList of FITS header values, stripping recognized keys if desired.
+    FitsSchemaInputMapper(daf::base::PropertyList & metadata, bool stripMetadata);
```

Comment:
The only code that could have passed `stripMetadata=false` was removed rather late in the refactor.  I'm not convinced the option needs to exist in the low-level code at all, but it is an option in all of our other FITS header parsing code.

I'm fairly ambivalent as to whether it's better to remove it to simplify the code a bit or leave it in for consistency with the other FITS header parsing code.  If you have a strong opinion, I'm happy to go with it.